File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/schedule/ScheduleFragment.java
package ru.sigil.fantasyradio.schedule;

import android.app.ProgressDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ExpandableListView;

import org.joda.time.LocalDate;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import ru.sigil.fantasyradio.R;
import ru.sigil.fantasyradio.dagger.Bootstrap;
import ru.sigil.log.LogManager;

public class ScheduleFragment extends Fragment {
    private String TAG = ScheduleFragment.class.getSimpleName();
    private ExpandableListView lv;
    private ParseAsyncTask searchAsyncTasc;
    ArrayList<ArrayList<ScheduleEntity>> arr = new ArrayList<>();
    private List<ScheduleEntity> scheduleEntityesCollection = new ArrayList<>();
    @Inject
    ScheduleParser scheduleParser;

    public ScheduleFragment() {
        super();
        setArguments(new Bundle());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        LogManager.d(TAG, "onCreateView");
        Bootstrap.INSTANCE.getBootstrap().inject(this);
        View scheduleFragmentView = inflater.inflate(R.layout.schedule_layout, container, false);
        scheduleFragmentView.findViewById(R.id.schedule_refresh_button).setOnClickListener(this::refreshClick);
        lv = scheduleFragmentView.findViewById(R.id.ScheduleListView);
        if (arr.size() > 0) {
            ScheduleListAdapter adapter = new ScheduleListAdapter(getActivity(), arr);
            lv.setAdapter(adapter);
            int count = adapter.getGroupCount();
            for (int position = 1; position <= count; position++)
                lv.expandGroup(position - 1);
        }
        return scheduleFragmentView;
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    /**
     * Парсим календарь в фоне и просим юзера подождать.
     */
    private class ParseAsyncTask extends AsyncTask<Void, Integer, Void> {

        ProgressDialog progress;

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            this.progress = new ProgressDialog(getActivity());
            this.progress.setMessage(getString(R.string.load));
            this.progress.setCancelable(false);
            this.progress.setOnCancelListener(dialog -> {
                // STOP thread execution
                searchAsyncTasc.cancel(true);
            });
            this.progress.show();
        }

        @Override
        protected Void doInBackground(Void... params) {
            scheduleEntityesCollection = scheduleParser.ParseSchedule();
            return null;
        }

        @Override
        protected void onProgressUpdate(Integer... progress) {
            super.onProgressUpdate(progress);
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
            arr = new ArrayList<>();
            for (int i = 0; i < 3; i++) {
                ArrayList<ScheduleEntity> arr2 = new ArrayList<>();
                LocalDate ld = LocalDate.now();
                ld = ld.plusDays(i);
                for (ScheduleEntity scheduleEntity : scheduleEntityesCollection) {
                    if (scheduleEntity.getStartDate() != null && scheduleEntity.getStartDate().getDayOfYear() == ld
                            .getDayOfYear()) {
                        arr2.add(scheduleEntity);
                    }
                }
                Collections.reverse(arr2);
                arr.add(arr2);
            }
            // ------------------------------------
            ScheduleListAdapter adapter = new ScheduleListAdapter(getActivity(), arr);
            lv.setAdapter(adapter);
            int count = adapter.getGroupCount();
            for (int position = 1; position <= count; position++)
                lv.expandGroup(position - 1);
            if (progress.isShowing()) {
                progress.dismiss();
            }
        }

        @Override
        protected void onCancelled() {
            if (progress.isShowing()) {
                progress.dismiss();
            }
        }
    }

    public void refreshClick(@SuppressWarnings("UnusedParameters") View v) {
        searchAsyncTasc = new ParseAsyncTask();
        searchAsyncTasc.execute();
    }

    @Override
    public void onResume() {
        LogManager.d(TAG, "onResume");
        super.onResume();
    }
}

File: FantasyRadio/src/main/kotlin/ru/sigil/fantasyradio/schedule/ScheduleFragment.kt
package ru.sigil.fantasyradio.schedule

import android.support.v4.app.Fragment
import javax.inject.Inject
import android.widget.ExpandableListView
import android.os.Bundle
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import ru.sigil.fantasyradio.R
import ru.sigil.fantasyradio.dagger.Bootstrap
import ru.sigil.log.LogManager
import android.app.ProgressDialog
import kotlinx.coroutines.*
import org.joda.time.LocalDate
import java.util.*

/**
 * Created by namelessone
 * on 02.12.18.
 */
class ScheduleFragment : Fragment() {
    init {
        arguments = Bundle()
    }

    private val TAG = ScheduleFragment::class.java.simpleName
    private var lv: ExpandableListView? = null
    private var arr = ArrayList<ArrayList<ScheduleEntity>>()
    private var scheduleEntitiesCollection: List<ScheduleEntity> = ArrayList()
    private var loadScheduleJob: Job? = null

    @Inject
    lateinit var scheduleParser: ScheduleParser

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        LogManager.d(TAG, "onCreateView")
        Bootstrap.INSTANCE.getBootstrap().inject(this)
        val scheduleFragmentView = inflater.inflate(R.layout.schedule_layout, container, false)
        scheduleFragmentView.findViewById<View>(R.id.schedule_refresh_button).setOnClickListener(this::refreshClick)
        lv = scheduleFragmentView.findViewById(R.id.ScheduleListView)
        if (arr.size > 0) {
            val adapter = ScheduleListAdapter(activity, arr)
            lv?.setAdapter(adapter)
            val count = adapter.groupCount
            for (position in 1..count)
                lv?.expandGroup(position - 1)
        }
        return scheduleFragmentView
    }

    /**
     * Парсим календарь в фоне и просим юзера подождать.
     */
    private fun refreshClick(v: View) {
        loadScheduleJob = GlobalScope.launch(Dispatchers.Main) {
            val progress = ProgressDialog(activity)
            try {
                //TODO timeout
                progress.setMessage(getString(R.string.load))
                progress.setCancelable(false)
                progress.show()
                GlobalScope.launch(Dispatchers.IO) {
                    scheduleEntitiesCollection = scheduleParser.parseSchedule()
                    arr = ArrayList()
                    for (i in 0..2) {
                        val arr2 = ArrayList<ScheduleEntity>()
                        var ld = LocalDate.now()
                        ld = ld.plusDays(i)
                        for (scheduleEntity in scheduleEntitiesCollection) {
                            if (scheduleEntity.startDate?.dayOfYear == ld.dayOfYear) {
                                arr2.add(scheduleEntity)
                            }
                        }
                        arr2.reverse()
                        arr.add(arr2)
                    }
                }.join()
                val adapter = ScheduleListAdapter(activity, arr)
                lv?.setAdapter(adapter)
                val count = adapter.groupCount
                for (position in 1..count)
                    lv?.expandGroup(position - 1)
            } finally {
                if (progress.isShowing) {
                    progress.dismiss()
                }
            }
        }
    }

    override fun onDestroyView() {
        loadScheduleJob?.cancel()
        super.onDestroyView()
    }
}